File: frontend/js/views/apr_apy.js
// APR ↔ APY — nominal vs effective annual rate at a compounding frequency,
// plus the continuous-compounding ceiling, via /calc/apr-apy. Updates live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const pct = (n) => Number(n).toFixed(4) + '%';

export async function renderAprApy(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.apy.h1.title">// APR ↔ APY</span></h1>
        <p class="muted small" data-i18n="view.apy.hint.intro">
            A rate quoted "12% compounded monthly" (APR) isn't what you actually earn or pay
            over a year — compounding pushes the effective rate (APY) higher, more so the
            oftener it compounds, up to a ceiling at continuous compounding. Convert either
            direction at any compounding frequency. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.apy.h2.inputs">The rate</h2>
            <form id="apy-form" class="inline-form">
                <label><span data-i18n="view.apy.label.direction">Convert</span>
                    <select name="direction">
                        <option value="apr_to_apy" data-i18n="view.apy.dir.apr_to_apy">APR → APY</option>
                        <option value="apy_to_apr" data-i18n="view.apy.dir.apy_to_apr">APY → APR</option>
                    </select></label>
                <label><span data-i18n="view.apy.label.rate">Rate (%)</span>
                    <input type="number" step="0.0001" min="0" name="rate_pct" value="12" required></label>
                <label><span data-i18n="view.apy.label.periods">Compounding periods / year</span>
                    <input type="number" step="1" min="1" name="periods_per_year" value="12" required></label>
            </form>
        </div>
        <div id="apy-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#apy-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            direction: fd.get('direction'),
            rate_pct: Number(fd.get('rate_pct')) || 0,
            periods_per_year: Number(fd.get('periods_per_year')) || 1,
        };
        try {
            const r = await api.calcAprApy(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.apy.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#apy-result');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.apy.h2.result">The rates</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.apy.card.apr">Nominal APR</div>
                    <div class="value">${pct(r.nominal_apr_pct)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.apy.card.apy">Effective APY</div>
                    <div class="value pos">${pct(r.effective_apy_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.apy.card.spread">Compounding boost</div>
                    <div class="value">${pct(r.spread_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.apy.card.continuous">Continuous-compounding APY</div>
                    <div class="value">${pct(r.continuous_apy_pct)}</div></div>
            </div>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/apr_apy.rs
//! APR ↔ APY — nominal vs effective annual rate.
//!
//! A nominal rate (APR) quoted "12% compounded monthly" isn't what you
//! actually earn or pay over a year — compounding makes the **effective**
//! annual rate (APY / EAR) higher. The more frequent the compounding, the
//! bigger the gap, up to a ceiling at continuous compounding:
//!
//!   * APY = (1 + APR/n)^n − 1
//!   * APR = n × ((1 + APY)^(1/n) − 1)
//!   * continuous-compounding APY = e^APR − 1
//!
//! Convert either direction given the compounding periods per year. Pure
//! compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Direction {
    /// Input is the nominal APR; compute the effective APY.
    AprToApy,
    /// Input is the effective APY; back out the nominal APR.
    ApyToApr,
}

#[derive(Debug, Clone, Deserialize)]
pub struct AprApyInput {
    pub direction: Direction,
    /// The input rate (percent) — interpreted per `direction`.
    pub rate_pct: f64,
    /// Compounding periods per year (1 annual, 12 monthly, 365 daily, …).
    pub periods_per_year: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct AprApyResult {
    pub nominal_apr_pct: f64,
    pub effective_apy_pct: f64,
    /// APY − APR: the compounding boost.
    pub spread_pct: f64,
    /// Effective rate if compounded continuously (e^APR − 1).
    pub continuous_apy_pct: f64,
}

pub fn analyze(i: &AprApyInput) -> AprApyResult {
    let n = if i.periods_per_year > 0.0 { i.periods_per_year } else { 1.0 };
    let rate = i.rate_pct / 100.0;

    let (apr, apy) = match i.direction {
        Direction::AprToApy => {
            let apy = (1.0 + rate / n).powf(n) - 1.0;
            (rate, apy)
        }
        Direction::ApyToApr => {
            let apr = n * ((1.0 + rate).powf(1.0 / n) - 1.0);
            (apr, rate)
        }
    };

    let continuous_apy = apr.exp() - 1.0;

    AprApyResult {
        nominal_apr_pct: apr * 100.0,
        effective_apy_pct: apy * 100.0,
        spread_pct: (apy - apr) * 100.0,
        continuous_apy_pct: continuous_apy * 100.0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn apr_to(rate: f64, n: f64) -> AprApyResult {
        analyze(&AprApyInput { direction: Direction::AprToApy, rate_pct: rate, periods_per_year: n })
    }

    #[test]
    fn monthly_compounding_apy() {
        // 12% APR monthly → (1.01)^12 − 1 = 12.6825%.
        let r = apr_to(12.0, 12.0);
        assert!((r.effective_apy_pct - 12.682503).abs() < 1e-4);
    }

    #[test]
    fn daily_compounding_apy_higher_than_monthly() {
        let monthly = apr_to(12.0, 12.0);
        let daily = apr_to(12.0, 365.0);
        assert!(daily.effective_apy_pct > monthly.effective_apy_pct);
        assert!((daily.effective_apy_pct - 12.7474).abs() < 1e-3);
    }

    #[test]
    fn annual_compounding_apy_equals_apr() {
        let r = apr_to(12.0, 1.0);
        assert!((r.effective_apy_pct - 12.0).abs() < 1e-9);
        assert!(r.spread_pct.abs() < 1e-9);
    }

    #[test]
    fn apy_to_apr_inverts_monthly() {
        // APY 12.682503% monthly → APR 12%.
        let r = analyze(&AprApyInput {
            direction: Direction::ApyToApr,
            rate_pct: 12.682503,
            periods_per_year: 12.0,
        });
        assert!((r.nominal_apr_pct - 12.0).abs() < 1e-3);
    }

    #[test]
    fn roundtrip_apr_apy_apr() {
        let fwd = apr_to(8.0, 12.0);
        let back = analyze(&AprApyInput {
            direction: Direction::ApyToApr,
            rate_pct: fwd.effective_apy_pct,
            periods_per_year: 12.0,
        });
        assert!((back.nominal_apr_pct - 8.0).abs() < 1e-6);
    }

    #[test]
    fn continuous_apy_is_exp_minus_one() {
        // e^0.12 − 1 = 12.7497%.
        let r = apr_to(12.0, 12.0);
        assert!((r.continuous_apy_pct - 12.749685).abs() < 1e-4);
    }

    #[test]
    fn spread_is_apy_minus_apr() {
        let r = apr_to(12.0, 12.0);
        assert!((r.spread_pct - (r.effective_apy_pct - r.nominal_apr_pct)).abs() < 1e-9);
        assert!(r.spread_pct > 0.0);
    }

    #[test]
    fn zero_rate_zero_everything() {
        let r = apr_to(0.0, 12.0);
        assert!(r.effective_apy_pct.abs() < 1e-9);
        assert!(r.continuous_apy_pct.abs() < 1e-9);
    }
}
